### PR TITLE
initial merge of INN work

### PR DESCRIPTION
**Pull request by aschweigert**
this is still minus the tests for API-dependent functions

### CODE REVIEW

**Comment by aschweigert** [comment]
@mike-douglas sorry this initial PR is so large, once this is merged the others should be much smaller/more discrete :)

**Comment by mike-douglas** [comment]
No worries!!! I spent some time looking over this today and doing some tests pulling and pushing content, and it looks great! I have some feedback, hopefully not big:

- Can we change the name of the plugin to "NPR Story API", and the author to "NPR Digital Services"?
- A few times I put in an invalid API key to see what kinds of errors I would get pushing a story, and aside from tailing the apache error logs, there is no output. Could we display a success/failure message at the top of the post page after I click the "Push to NPR" button?
- Could we change the title of the "Get DS NPR Stories" page (under Posts) to simply "Get NPR Stories"?
- I see that the NPR story ID is added to the Post when it is successfully pushed to the API, but when I look at the story again I don't see it in custom fields. Could you add that? It would be nice if it showed up as a custom field, like the way that the pulled stories do. It helps with debugging with stations.

**Comment by aschweigert** [comment]
@mike-douglas sure, no problem, the first three items should be no big deal, we'll take a look at those today or tomorrow. The last item might be a bit more problematic because we'd need to rename that custom field so WP doesn't hide it from regular users. That *might* be ok, but it might also require a migration script for folks already using the plugin, some other changes elsewhere in the plugin, etc. so that's probably a bit more involved. Definitely possible though so let me know if that's needed and we'll figure it out. We're still doing well on time/budget here so even if that takes a few hours to work out I think it's doable.

**Comment by benlk** [comment]
- [x] Change the name of the plugin to "NPR Story API", and the author to "NPR Digital Services"
- [x] Change the title of the "Get DS NPR Stories" page (under Posts) to simply "Get NPR Stories"
- [ ] Could we display a success/failure message at the top of the post page after I click the "Push to NPR" button?
    - [x] failure message
    - [ ] success message
- [ ] add NPR story ID to stories that have been pushed from the plugin.

**Comment by mike-douglas** [comment]
@aschweigert it might be easier to add a success notification with the Story ID if you can't store it. Ideally we'd have it saved somewhere so we could query for the story that was posted in the API, but this might be good enough.

**Comment by benlk** [comment]
@mike-douglas Pressing the "Push to NPR" button has triggered a "Post Updated" message on the post in earlier versions of this pull request. This was apparently a default WordPress behavior. https://github.com/nprds/nprapi-wordpress/pull/19/commits/179fe2c5083c5f53046a90f0146ba056fe8a9c9a extends that message to include the NPR Story ID.

The story ID itself is being saved as the `npr_story_id` for both pushed posts and pulled posts, and is showing up in the Custom Fields view for both sorts of posts in my testing. Were you not seeing it before?

**Comment by benlk** [comment]
Here's the results of the initial submission to wordpress.org for review:

> There are issues with your plugin code. Please read this ENTIRE email, address all listed issues, and reply to this email with your corrected code attached. It is required for you to read and reply to these emails, and failure to do so will result in significant delays with your plugin being accepted. 
> 
> ## Generic function (and/or define) names 
> 
> All plugins should have unique function names, defines, and classnames. This will prevent your plugin from conflicting with other plugins or themes. 
> 
> For example, if your plugin is called "Easy Custom Post Types", then you might prefix your functions with ecpt_{your function name here}. Similarly a define of `LICENSE` would be better done as `ECPT_LICENSE`. You can use namespaces instead, however make sure that those also are unique. A namespace or class of 'MyPlugin' is NOT actually all that unique, you see. 
> This extends to anything in a define. For example ... 
> 
>     define(` 'PLUGIN_PATH', plugins_url( __FILE__ ) ); 
> 
> That define is a global, so `PLUGIN_PATH` could conflict with a number of other things. 
> 
> Don't try to use two letter slugs anymore. As of 2016, all the good ones are taken. Instead consider `easy_cpts_` (from the first example). 
> 
> Similarly, don't use `__` to prefix, as the double underscore should be reserved for WordPress itself. 
> 
> Please update your plugin to use more unique names. 
> 
> function `npr_push` is PROBABLY not a great name, since you may make other npr plugins. This is a case where `npr_story_push` would possibly work better... Though you use `ds_npr_` in most places, so maybe just being consistent would be enough there. 
> 
> `as_nprml` is fairly risky in so far as prefixes go. 
> 
> Just FYI, function `_ds_npr_activate()` is not a good name `_`'s should be left alone for parent projects, or for use inside singletons. Causes less stupid. 
> 
> ## Calling core loading files directly 
> 
> Including `wp-config.php`, `wp-blog-header.php`, `wp-load.php`, or pretty much any other WordPress core file that you have to call directly via an include is not a good idea and we cannot approve a plugin that does so unless it has a very good reason to load the file(s). It is prone to failure since not all WordPress installs have the exact same file structure. 
> 
> Usually plugins will include `wp-config.php` or `wp-load.php` in order to gain access to core WordPress functions, but there are much better ways to do this. It's best if you tie your processing functions (the ones that need but don't have access to core functions) into an action hook, such as `init` or `admin_init`. 
> 
> Please consult the Plugins API reference for more information: http://codex.wordpress.org/Plugin_API 
> 
> If you're trying to use AJAX, please read this: http://codex.wordpress.org/AJAX_in_Plugins 
> 
> For other possibilities, or to better understand why we disallow this, read this: http://ottopress.com/2010/dont-include-wp-load-please/ 
> 
> If you're trying to use it because you need to access WordPress functions outside of WordPress, we'd actually much rather you didn't do that at all. Your plugin should be inside WordPress, only accessible to people who are logged in and authorized, if it needs that kind of access. Your plugin's pages should be called via the dashboard like all the other settings panels, and in that way, they'll always have access to WordPress functions. 
> 
> ## Calling file locations poorly 
> 
> The way your plugin is referencing other files is not going to work with all setups of WordPress. 
> 
>     require_once (WP_PLUGIN_DIR . '/../../wp-admin/includes/file.php'); 
>     require_once (WP_PLUGIN_DIR . '/../../wp-admin/includes/media.php'); 
>     require_once (WP_PLUGIN_DIR . '/../../wp-admin/includes/admin.php'); 
> 
> 
> When you hardcode in paths, or assume that everyone has WordPress in the root of their domain, you cause anyone using 'Giving WordPress it's own directory' (a VERY common setup) to break. In addition, WordPress allows users to change the name of wp-content, so you would break anyone who choses to do so. 
> 
> Please review http://codex.wordpress.org/Determining_Plugin_and_Content_Directories and update your plugin accordingly. And don't worry about supporting WordPress 2.x or lower. We don't encourage it nor expect you to do so, so save yourself some time and energy. 
> 
> You'll want `ABSPATH . '/wp-admin...'` and so on for those files. 
> 
> ## Please sanitize, escape, and validate your POST calls 
> 
> When you include POST/GET/REQUEST calls in your plugin, it's important to sanitize, validate, and escape them. The goal here is to prevent a user from accidentally sending trash data through the system, as well as protecting them from potential security issues. 
> 
> SANITIZE: All instances where generated content is inserted into the database, or into a file, or being otherwise processed by WordPress, the data MUST be properly sanitized for security. By sanitizing your POST data when used to make action calls or URL redirects, you will lessen the possibility of XSS vulnerabilities. You should never have a raw data inserted into the database, even by a update function, and even with a `prepare()` call. 
> 
> VALIDATE: In addition to sanitization, you should validate all your calls. If a `$_POST` call should only be a number, ensure it's an `int()` before you pass it through anything. Even if you're sanitizing or using WordPress functions to ensure things are safe, we ask you please validate for sanity's sake. Any time you are adding data to the database, it should be the right data. 
> 
> ESCAPE: Similarly, when you're outputting data, make sure to escape it properly, so it can't hijack admin screens. There are many `esc_*()` functions you can use to make sure you don't show people the wrong data. 
> 
> In all cases, using `stripslashes` or `strip_tags` is not enough. You need to use the most appropriate method associated with the type of content you're processing. Check that a URL is a URL and don't just be lazy and use `sanitize_text` please. The ultimate goal is that you should ensure that invalid and unsafe data is NEVER processed or displayed. Clean everything, check everything, escape everything, and never trust the users to always have input sane data. 
> 
> Please review this document and update your code accordingly: http://codex.wordpress.org/Validating_Sanitizing_and_Escaping_User_Data 
> 
> Example: 
> 
>     $story_id = $_POST[ 'story_id' ] ; 
> 
> You should `intval` check that, save a small headache when someone tries to pass through junk data. 
> 
> ## Not using Nonces and/or checking permissions 
> 
> Please add a nonce to your POST calls to prevent unauthorized access. 
> Keep in mind, `check_admin_referer` alone is NOT bulletproof security. Do not rely on nonces for authorization purposes. Use `current_user_can()` in order to prevent users without the right permissions from accessing things. 
> 
> https://codex.wordpress.org/WordPress_Nonces 
> 
> 
> ---- 
> 
> Please make sure you've addressed ALL issues brought up in this email. When you've corrected your code, reply to this email with the updated code attached as a zip, or provide a link to the new code for us to review. If you have questions, concerns, or need clarification, please reply to this email and just ask us. 
> 
> (While we have tried to make this review as exhaustive as possible we, like you, are humans and may have missed things. As such, we will re-review the ENTIRE plugin when you send it back to us. We appreciate your patience and understanding in this.)

So:

- [ ] change namespacing to use ds_npr across the board, or something else
- [ ] change name of as_nprml, unless there's a very strong reason to not do so
- [ ] fing the WP core files being loaded directly and fix that, possibly rewriting the wp cron task using them or appealing that. It depends on how cron tasks should be written.
- [ ] fix paths
- [ ] add nonces
- [ ] sanitization/validation

Once all that is done in this PR, we'll generate another .zip and email that to y'all to email back to them from the email address associated with the `nprds` wordpress.org account.

**Comment by mike-douglas** [comment]
@benlk I did see the ID stored there in the database, but I didn't see it listed as a custom field in the interface. When I pull a story from the API, I can see all of the additional fields and I was hoping to see the same with the story ID for local stories that are pushed

**Comment by benlk** [comment]
It's curious that you're not seeing the ID in the custom fields interface when you push a story. All the stories I pushed to https://api-s1.npr.org yesterday displayed the ID in the custom fields:
 
screen shot 2016-06-22 at 10 44 23 am

**Comment by mike-douglas** [comment]
@benlk My bad, the custom fields were not showing with my quickly set-up install of wordpress. I can confirm that they're showing now 😅

File: get_stories.php [hunk 1]
```diff
@@ -18,13 +18,14 @@ public static function nprstory_cron_pull() {
 		//because this is run on cron, and may be fired off by an non-admin, we need to load a bunch of stuff
 		require_once (ABSPATH . '/wp-admin/includes/file.php');
 		require_once (ABSPATH . '/wp-admin/includes/media.php');
-		require_once (ABSPATH . '/wp-admin/includes/admin.php');
-		require_once (ABSPATH . '/wp-includes/class-wp-error.php');
```
Comment (benlk) [inline get_stories.php]: `/wp-cron.php` [loads](https://core.trac.wordpress.org/browser/tags/4.5.3/src/wp-cron.php#L26) `/wp-load.php`
`/wp-load.php` either [loads `/wp-settings.php` directly](https://core.trac.wordpress.org/browser/tags/4.5.3/src/wp-load.php#L39) or loads `wp-config.php`which loads `/wp-settings.php` - in either case  `/wp-settings.php` is loaded
 `/wp-settings.php` [loads](https://core.trac.wordpress.org/browser/tags/4.5.3/src/wp-settings.php#L83) `WPINC . '/class-wp-error.php'`

Thus we don't need to load `/wp-includes/class-wp-error.php`

`/wp-admin/includes/admin.php` was not required in testing of this section.

`/wp-admin/includes/file.php` is required to access the API.

`/wp-admin/includes/media.php` is required for image handling.

**Comment by benlk** [comment]
Left to do before the wp.org re-submit:

- [ ] nonces
- [ ] sanitization
- [ ] validation

**Comment by benlk** [comment]
- [ ] check nonces

**Comment by benlk** [comment]
This is now in a state that we should resubmit it to wordpress.org.

- require certain WordPress capabilities before performing API actions:
    - deleting posts in the API now requires the delete_others_posts capability for the user trying to delete the post. In effect, this limits deletion of stories from the NPR API to admins and editors.
    - pushing stories to the API requires the publish_posts capability, which means that contributors and guests can't push to NPR.
    - pulling posts from the API requires the edit_posts capability, meaning that contributors and guests can't pull from NPR. We may want to revisit this in the future, or make the permissions filterable if sites using the plugin want to lock permissions down or open them up.
- stronger validation inside `DS_NPR_API::load_page_hook`: https://github.com/INN/WP-DS-NPR-API/compare/1069e4982bde68faf75dad03a88a21e96f76e1f3...18a07aadae59bc4574eb7d154f15cf29b3bb95d3#diff-81d75f48e65ff14ba06969d4d2235995R72
- all functions not inside classes have been renamed with a `nprstory_` prefix - this is the bulk of the pull request's line diffs
- The defined constants are all using `NPR_` or `NPRSTORY_`, so in the future we may want to update the `NPR_` ones to `NPRSTORY_` constant names.
- Most settings still use the `ds_npr_` prefix. I didn't change those because we'd need to write a migration function for all the functions, and wp.org didn't complain about the settings. We might want to rename them later, but that's not essential.
- found bugs: https://github.com/nprds/nprapi-wordpress/issues/20 https://github.com/nprds/nprapi-wordpress/issues/21
- redid how a bunch of files were `require_once`'d
- removed `/wp-admin/includes/admin.php`, `/wp-load.php`, `/wp-includes/class-wp-error.php` from the cron job function, see https://github.com/nprds/nprapi-wordpress/pull/19/commits/4b3d65a19122b0da5215997939db94c7accf3e5e and https://github.com/nprds/nprapi-wordpress/pull/19/commits/cf2dfa39c1f118d0ca0c836d7967e09baec63bd6 - the cron job works without them.
- update constructor for `DS_NPR_API` class
- use the `number` input type for a couple of settings in the admin
- added nonces for all plugin settings
- add sanitization/validation callbacks for all of the plugin's settings - these could be a lot more restrictive than they currently are, though. Most do not check nonces.

**Comment by mike-douglas** [comment]
I resubmitted at around 5pm, awaiting a response.. ⏳

**Pull request merged**